internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,468 @@
+					continue
+				}
+
+				if shouldSwitch {
+					if arch, ok := cpuArchMapFromAPIArch[cpuArchitecture]; ok {
+						cpuArchitecture = arch
+					}
+				}
+
+				for _, node := range graph.Nodes {
+					if strings.HasPrefix(node.Version, openshiftVersion) {
+						// Got a match for OpenShiftReleaseImage
+						releases, err = h.addReleaseImage(releases, channel, node.Version, cpuArchitecture)
+						if err != nil {
+							return nil, err
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// remove duplicates
+
+	h.log.Debugf("Got %d releases", len(releases))
+
+	return releases, nil
+}
+
+func (h *ReleaseSourcesHandler) getConfigurationReleaseImages() []*common.ReleaseImage {
+	dbReleaseImages := []*common.ReleaseImage{}
+
+	for _, releaseImage := range h.releaseImages {
+		dbReleaseImage := common.ReleaseImage{ReleaseImage: *releaseImage}
+		dbReleaseImage.CPUArchitectures = nil
+		dbReleaseImages = append(dbReleaseImages, &dbReleaseImage)
+	}
+
+	return dbReleaseImages
+}
+
+// mergeReleaseImages merges configuration and db release images together such that
+// configuration release images have precedence in case of conflicts e.g. which release is default
+func (h *ReleaseSourcesHandler) mergeReleaseImages(configurationReleaseImages, apiReleaseImages []*common.ReleaseImage) []*common.ReleaseImage {
+	mergedReleaseImages := []*common.ReleaseImage{}
+	configurationReleaseImagesReferenceSet := map[string]bool{}
+	configurationHasDefaultRelease := false
+
+	for _, configurationRelease := range configurationReleaseImages {
+		mergedReleaseImages = append(mergedReleaseImages, configurationRelease)
+		configurationReleaseImagesReferenceSet[*configurationRelease.URL] = true
+		if configurationRelease.Default {
+			configurationHasDefaultRelease = true
+		}
+	}
+
+	for _, apiRelease := range apiReleaseImages {
+		if configurationReleaseImagesReferenceSet[*apiRelease.URL] {
+			continue
+		}
+
+		if apiRelease.Default && configurationHasDefaultRelease {
+			apiRelease.Default = false
+		}
+
+		mergedReleaseImages = append(mergedReleaseImages, apiRelease)
+	}
+
+	return mergedReleaseImages
+}
+
+func (h *ReleaseSourcesHandler) syncReleaseImages(tx *gorm.DB) error {
+	apiReleaseImages, err := h.getReleaseImagesFromAPI()
+	if err != nil {
+		return err
+	}
+	err = h.setDefaultRelease(apiReleaseImages)
+	if err != nil {
+		return err
+	}
+	supportLevels, err := h.syncOpenshiftVersionsSupportLevels(tx)
+	if err != nil {
+		return err
+	}
+	err = h.setSupportLevels(apiReleaseImages, supportLevels)
+	if err != nil {
+		return err
+	}
+	configurationReleaseImages := h.getConfigurationReleaseImages()
+	err = h.setSupportLevels(configurationReleaseImages, supportLevels)
+	if err != nil {
+		return err
+	}
+	apiReleaseImages = h.removeReleaseImageDuplicates(apiReleaseImages)
+	releaseImages := h.mergeReleaseImages(configurationReleaseImages, apiReleaseImages)
```

Comment:
Supporting this seems like opening a can of worms, what's the use case of merging API release images with user configuration release images?

Reply:
Can't we just get all the release images from the `RELEASE_SOURCES` API? Why do we still need to maintain a config if we have API access?

I understand the use case for `RELEASE_SOURCES` in isolation, I understand the user case for `RELEASE_IMAGES` in isolation, but I'm not sure why would anyone would want to have both of them together and why would we support that

Reply:
Because using `RELEASE_SOURCES` we can fetch releases according to `version`, `channel` and `cpu_architecture` - so for example if a client have a special approval to use an old release which we don't support anymore, lets say 4.10.<something> (x86_64), then in order to provide this release using only `RELEASE_SOURCES` we would have to fetch all (stable 4.9 x86_64) from the API, and keep them in the DB only for one release. The other alternative to use only `RELEASE_IMAGE` requires maintaining log configuration lists as we do now but much longer